src/main/java/com/nike/cerberus/endpoints/secret/ReadSecureData.java

```diff
@@ -0,0 +1,138 @@
+import com.nike.riposte.util.AsyncNettyHelper;
+import com.nike.riposte.util.Matcher;
+import com.nike.riposte.util.MultiMatcher;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import javax.ws.rs.core.SecurityContext;
+import java.io.IOException;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+public class ReadSecureData extends SecureDataEndpointV1<Void, Object> {
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    @Inject
+    protected ReadSecureData(SecureDataService secureDataService,
+                             PermissionsService permissionService,
+                             SafeDepositBoxService safeDepositBoxService) {
+
+        super(secureDataService, permissionService, safeDepositBoxService);
+    }
+
+    @Override
+    protected ResponseInfo<Object> generateVaultStyleResponse(VaultStyleErrorResponse response, int statusCode) {
+        return ResponseInfo.newBuilder()
+                .withContentForFullResponse(response)
+                .withHttpStatusCode(statusCode)
+                .build();
+    }
+
+    @Override
+    public CompletableFuture<ResponseInfo<Object>> doExecute(SecureDataRequestInfo requestInfo,
+                                                                         RequestInfo<Void> request,
+                                                                         Executor longRunningTaskExecutor,
+                                                                         ChannelHandlerContext ctx,
+                                                                         SecurityContext securityContext) {
+
+        if (StringUtils.equalsIgnoreCase(request.getQueryParamSingle("list"), "true")) {
+            return CompletableFuture.supplyAsync(
+                    AsyncNettyHelper.supplierWithTracingAndMdc(() -> listKeys(requestInfo), ctx),
+                    longRunningTaskExecutor
+            );
+        }
+
+        return CompletableFuture.supplyAsync(
+                AsyncNettyHelper.supplierWithTracingAndMdc(() -> readSecureData(requestInfo), ctx),
+                longRunningTaskExecutor
+        );
+    }
+
+    private ResponseInfo<Object> listKeys(SecureDataRequestInfo info) {
+        Set<String> keys = secureDataService.listKeys(info.getPath());
+
+        if (keys.isEmpty()) {
+            return generateVaultStyleResponse(new VaultStyleErrorResponse.VaultStyleErrorResponseBuilder().build(),
+                    HttpResponseStatus.NOT_FOUND.code());
+        }
+
+        SecureDataResponse response = new SecureDataResponse();
+        response.setRequestId(UUID.randomUUID().toString()); // maybe use trace id?
```

Comment:
This seems weird.  Yes, maybe trace id so we can correlate it to something?

Reply:
- [ ] Use trace id